File: by-language/python/data_structures/queue.py
"""
Queue (FIFO - First In, First Out)

╔════════════════════════════════════════════════════════════════════════════╗
║                           VISUAL REPRESENTATION                            ║
╠════════════════════════════════════════════════════════════════════════════╣
║                                                                            ║
║  Queue Structure:                                                          ║
║  ┌─────┬─────┬─────┬─────┬─────┐                                           ║
║  │  10 │  20 │  30 │  40 │  50 │                                           ║
║  └─────┴─────┴─────┴─────┴─────┘                                           ║
║    ▲                         ▲                                             ║
║   front                    rear                                            ║
║  (dequeue)               (enqueue)                                         ║
║                                                                            ║
║  Enqueue Operation (add to rear):                                          ║
║  Before: [10, 20, 30]                                                      ║
║  enqueue(40)                                                               ║
║  After:  [10, 20, 30, 40]                                                  ║
║                                                                            ║
║  Dequeue Operation (remove from front):                                    ║
║  Before: [10, 20, 30, 40]                                                  ║
║  dequeue() → returns 10                                                    ║
║  After:  [20, 30, 40]                                                      ║
║                                                                            ║
║  Circular Queue (size=4):                                                  ║
║  ┌─────┬─────┬─────┬─────┐                                                 ║
║  │  30 │  40 │     │  20 │                                                 ║
║  └─────┴─────┴─────┴─────┘                                                 ║
║    [0]   [1]   [2]   [3]                                                   ║
║                 ▲     ▲                                                    ║
║               rear  front                                                  ║
║                                                                            ║
║  Deque (Double-Ended Queue):                                               ║
║  ┌─────┬─────┬─────┬─────┬─────┐                                           ║
║  │  10 │  20 │  30 │  40 │  50 │                                           ║
║  └─────┴─────┴─────┴─────┴─────┘                                           ║
║    ▲                         ▲                                             ║
║   front                    rear                                            ║
║  (add/remove)          (add/remove)                                        ║
║                                                                            ║
╚════════════════════════════════════════════════════════════════════════════╝

COMPLEXITY:
┌─────────────────────┬─────────────┬─────────────┬─────────────┐
│ Operation           │ Time        │ Space       │ Notes       │
├─────────────────────┼─────────────┼─────────────┼─────────────┤
│ enqueue             │ O(1)*       │ O(1)        │ *amortized  │
│ dequeue             │ O(1)        │ O(1)        │             │
│ peek/front          │ O(1)        │ O(1)        │             │
│ is_empty            │ O(1)        │ O(1)        │             │
│ size                │ O(1)        │ O(1)        │             │
│ Deque add/remove    │ O(1)        │ O(1)        │ both ends   │
└─────────────────────┴─────────────┴─────────────┴─────────────┘

LEETCODE PROBLEMS:
- #232 Implement Queue using Stacks
- #225 Implement Stack using Queues
- #622 Design Circular Queue
- #641 Design Circular Deque
- #346 Moving Average from Data Stream
- #933 Number of Recent Calls
- #239 Sliding Window Maximum (deque)
- #862 Shortest Subarray with Sum at Least K
- #1438 Longest Continuous Subarray With Absolute Diff

USE CASES:
- Task scheduling (CPU, printer queue)
- Breadth-first search (BFS)
- Buffer for data streams
- Handling requests in order
- Message queues
- Keyboard buffer
- Cache implementation
- Level-order tree traversal
"""

from typing import TypeVar, Generic, Iterator, Optional, List
from collections import deque

T = TypeVar('T')


class Queue(Generic[T]):
    """
    A queue implementation using Python's deque for O(1) operations.

    Uses collections.deque which provides O(1) append and popleft.
    """

    def __init__(self) -> None:
        """Initialize an empty queue."""
        self._data: deque[T] = deque()

    def __len__(self) -> int:
        """Return the number of elements in the queue."""
        return len(self._data)

    def __iter__(self) -> Iterator[T]:
        """Iterate from front to rear."""
        return iter(self._data)

    def __repr__(self) -> str:
        """Return string representation."""
        return f"Queue({list(self._data)})"

    def __eq__(self, other: object) -> bool:
        """Check equality with another queue."""
        if not isinstance(other, Queue):
            return NotImplemented
        return list(self._data) == list(other._data)

    def __contains__(self, value: T) -> bool:
        """Check if value exists in queue."""
        return value in self._data

    @property
    def is_empty(self) -> bool:
        """Return True if queue is empty."""
        return len(self._data) == 0

    def enqueue(self, value: T) -> None:
        """
        Add an element to the rear of the queue.

        Time Complexity: O(1)

        Args:
            value: The value to add.
        """
        self._data.append(value)

    def dequeue(self) -> T:
        """
        Remove and return the front element.

        Time Complexity: O(1)

        Returns:
            The front element.

        Raises:
            IndexError: If the queue is empty.
        """
        if self.is_empty:
            raise IndexError("Dequeue from empty queue")
        return self._data.popleft()

    def peek(self) -> T:
        """
        Return the front element without removing it.

        Time Complexity: O(1)

        Returns:
            The front element.

        Raises:
            IndexError: If the queue is empty.
        """
        if self.is_empty:
            raise IndexError("Peek from empty queue")
        return self._data[0]

    def rear(self) -> T:
        """
        Return the rear element without removing it.

        Time Complexity: O(1)

        Returns:
            The rear element.

        Raises:
            IndexError: If the queue is empty.
        """
        if self.is_empty:
            raise IndexError("Rear from empty queue")
        return self._data[-1]

    def clear(self) -> None:
        """Remove all elements from the queue."""
        self._data.clear()

    def to_list(self) -> List[T]:
        """Convert queue to list (front to rear)."""
        return list(self._data)

    @classmethod
    def from_list(cls, items: List[T]) -> "Queue[T]":
        """
        Create a queue from a list.

        First element becomes front of queue.

        Args:
            items: List of items.

        Returns:
            A new Queue.
        """
        queue: Queue[T] = cls()
        for item in items:
            queue.enqueue(item)
        return queue


class CircularQueue(Generic[T]):
    """
    A fixed-size circular queue implementation using an array.

    Efficient use of space by wrapping around when reaching the end.
    """

    def __init__(self, capacity: int) -> None:
        """
        Initialize a circular queue with fixed capacity.

        Args:
            capacity: Maximum number of elements.

        Raises:
            ValueError: If capacity is less than 1.
        """
        if capacity < 1:
            raise ValueError("Capacity must be at least 1")

        self._capacity: int = capacity
        self._data: List[Optional[T]] = [None] * capacity
        self._front: int = 0
        self._rear: int = -1
        self._size: int = 0

    def __len__(self) -> int:
        """Return the number of elements."""
        return self._size

    def __repr__(self) -> str:
        """Return string representation."""
        elements = []
        if self._size > 0:
            idx = self._front
            for _ in range(self._size):
                elements.append(str(self._data[idx]))
                idx = (idx + 1) % self._capacity
        return f"CircularQueue([{', '.join(elements)}], capacity={self._capacity})"

    @property
    def is_empty(self) -> bool:
        """Return True if queue is empty."""
        return self._size == 0

    @property
    def is_full(self) -> bool:
        """Return True if queue is full."""
        return self._size == self._capacity

    @property
    def capacity(self) -> int:
        """Return the capacity of the queue."""
        return self._capacity

    def enqueue(self, value: T) -> bool:
        """
        Add an element to the rear.

        Time Complexity: O(1)

        Args:
            value: The value to add.

        Returns:
            True if successful, False if queue is full.
        """
        if self.is_full:
            return False

        self._rear = (self._rear + 1) % self._capacity
        self._data[self._rear] = value
        self._size += 1
        return True

    def dequeue(self) -> Optional[T]:
        """
        Remove and return the front element.

        Time Complexity: O(1)

        Returns:
            The front element, or None if empty.
        """
        if self.is_empty:
            return None

        value = self._data[self._front]
        self._data[self._front] = None
        self._front = (self._front + 1) % self._capacity
        self._size -= 1
        return value

    def peek(self) -> Optional[T]:
        """
        Return the front element without removing it.

        Time Complexity: O(1)

        Returns:
            The front element, or None if empty.
        """
        if self.is_empty:
            return None
        return self._data[self._front]

    def rear(self) -> Optional[T]:
        """
        Return the rear element without removing it.

        Time Complexity: O(1)

        Returns:
            The rear element, or None if empty.
        """
        if self.is_empty:
            return None
        return self._data[self._rear]


class Deque(Generic[T]):
    """
    A double-ended queue (deque) implementation.

    Supports O(1) insertion and deletion at both ends.
    """

    def __init__(self) -> None:
        """Initialize an empty deque."""
        self._data: deque[T] = deque()

    def __len__(self) -> int:
        """Return the number of elements."""
        return len(self._data)

    def __iter__(self) -> Iterator[T]:
        """Iterate from front to rear."""
        return iter(self._data)

    def __repr__(self) -> str:
        """Return string representation."""
        return f"Deque({list(self._data)})"

    def __getitem__(self, index: int) -> T:
        """Get element at index."""
        return self._data[index]

    @property
    def is_empty(self) -> bool:
        """Return True if deque is empty."""
        return len(self._data) == 0

    def add_front(self, value: T) -> None:
        """
        Add element to the front.

        Time Complexity: O(1)

        Args:
            value: The value to add.
        """
        self._data.appendleft(value)

    def add_rear(self, value: T) -> None:
        """
        Add element to the rear.

        Time Complexity: O(1)

        Args:
            value: The value to add.
        """
        self._data.append(value)

    def remove_front(self) -> T:
        """
        Remove and return the front element.

        Time Complexity: O(1)

        Returns:
            The front element.

        Raises:
            IndexError: If deque is empty.
        """
        if self.is_empty:
            raise IndexError("Remove from empty deque")
        return self._data.popleft()

    def remove_rear(self) -> T:
        """
        Remove and return the rear element.

        Time Complexity: O(1)

        Returns:
            The rear element.

        Raises:
            IndexError: If deque is empty.
        """
        if self.is_empty:
            raise IndexError("Remove from empty deque")
        return self._data.pop()

    def peek_front(self) -> T:
        """
        Return the front element without removing.

        Time Complexity: O(1)

        Returns:
            The front element.

        Raises:
            IndexError: If deque is empty.
        """
        if self.is_empty:
            raise IndexError("Peek from empty deque")
        return self._data[0]

    def peek_rear(self) -> T:
        """
        Return the rear element without removing.

        Time Complexity: O(1)

        Returns:
            The rear element.

        Raises:
            IndexError: If deque is empty.
        """
        if self.is_empty:
            raise IndexError("Peek from empty deque")
        return self._data[-1]

    def clear(self) -> None:
        """Remove all elements."""
        self._data.clear()

    def to_list(self) -> List[T]:
        """Convert to list."""
        return list(self._data)


class MonotonicQueue(Generic[T]):
    """
    A monotonic deque that maintains elements in sorted order.

    Useful for sliding window maximum/minimum problems.
    This implementation maintains a decreasing order (for maximum queries).
    """

    def __init__(self) -> None:
        """Initialize an empty monotonic queue."""
        self._data: deque[tuple[int, T]] = deque()  # (index, value)

    def __len__(self) -> int:
        """Return the number of elements."""
        return len(self._data)

    @property
    def is_empty(self) -> bool:
        """Return True if empty."""
        return len(self._data) == 0

    def push(self, index: int, value: T) -> None:
        """
        Add element maintaining decreasing order.

        Removes all elements smaller than the new value.

        Time Complexity: O(1) amortized

        Args:
            index: The index of the element.
            value: The value to add.
        """
        while self._data and self._data[-1][1] < value:  # type: ignore
            self._data.pop()
        self._data.append((index, value))

    def pop(self, index: int) -> None:
        """
        Remove element at index if it's at the front.

        Time Complexity: O(1)

        Args:
            index: The index to remove.
        """
        if self._data and self._data[0][0] == index:
            self._data.popleft()

    def get_max(self) -> Optional[T]:
        """
        Get the maximum element.

        Time Complexity: O(1)

        Returns:
            The maximum element, or None if empty.
        """
        if self.is_empty:
            return None
        return self._data[0][1]


def sliding_window_maximum(nums: List[int], k: int) -> List[int]:
    """
    Find maximum in each sliding window of size k.

    Time Complexity: O(n)
    Space Complexity: O(k)

    Args:
        nums: List of integers.
        k: Window size.

    Returns:
        List of maximums for each window position.

    Example:
        nums = [1,3,-1,-3,5,3,6,7], k = 3
        Output: [3,3,5,5,6,7]
    """
    if not nums or k <= 0:
        return []

    if k == 1:
        return nums.copy()

    result: List[int] = []
    mq = MonotonicQueue[int]()

    for i, num in enumerate(nums):
        mq.push(i, num)

        # Remove elements outside the window
        if i >= k:
            mq.pop(i - k)

        # Add maximum to result once window is formed
        if i >= k - 1:
            max_val = mq.get_max()
            if max_val is not None:
                result.append(max_val)

    return result

File: by-language/rust/dsa-data-structures/src/linear/queue.rs
//! Queue (FIFO - First In, First Out)
//!
//! A queue is a linear data structure that follows the FIFO principle.
//!
//! ```text
//! ╔════════════════════════════════════════════════════════════════════════════╗
//! ║                           VISUAL REPRESENTATION                            ║
//! ╠════════════════════════════════════════════════════════════════════════════╣
//! ║                                                                            ║
//! ║  Queue Structure:                                                          ║
//! ║  ┌─────┬─────┬─────┬─────┬─────┐                                           ║
//! ║  │  1  │  2  │  3  │  4  │  5  │                                           ║
//! ║  └─────┴─────┴─────┴─────┴─────┘                                           ║
//! ║    ▲                       ▲                                               ║
//! ║  front                   rear                                              ║
//! ║  (dequeue)             (enqueue)                                           ║
//! ║                                                                            ║
//! ║  Enqueue Operation:                                                        ║
//! ║  ┌─────┬─────┬─────┬─────┬─────┬─────┐                                     ║
//! ║  │  1  │  2  │  3  │  4  │  5  │ NEW │ ◀── enqueue(6)                       ║
//! ║  └─────┴─────┴─────┴─────┴─────┴─────┘                                     ║
//! ║                                                                            ║
//! ║  Dequeue Operation:                                                        ║
//! ║  ┌─────┬─────┬─────┬─────┬─────┐                                           ║
//! ║  │  X  │  2  │  3  │  4  │  5  │                                           ║
//! ║  └─────┴─────┴─────┴─────┴─────┘                                           ║
//! ║    ▲                                                                       ║
//! ║  dequeue() returns 1, new front is 2                                       ║
//! ║                                                                            ║
//! ╚════════════════════════════════════════════════════════════════════════════╝
//! ```
//!
//! ## Complexity
//!
//! | Operation | Average | Worst | Space |
//! |-----------|---------|-------|-------|
//! | Enqueue   | O(1)*   | O(n)  | O(1)  |
//! | Dequeue   | O(1)*   | O(n)  | O(1)  |
//! | Peek      | O(1)    | O(1)  | O(1)  |
//! | Search    | O(n)    | O(n)  | O(1)  |
//! | Is Empty  | O(1)    | O(1)  | O(1)  |
//!
//! *Amortized O(1) using VecDeque
//!
//! ## LeetCode Problems
//!
//! - [#225 Implement Stack using Queues](https://leetcode.com/problems/implement-stack-using-queues/)
//! - [#622 Design Circular Queue](https://leetcode.com/problems/design-circular-queue/)
//! - [#933 Number of Recent Calls](https://leetcode.com/problems/number-of-recent-calls/)
//! - [#346 Moving Average from Data Stream](https://leetcode.com/problems/moving-average-from-data-stream/)
//! - [#102 Binary Tree Level Order Traversal](https://leetcode.com/problems/binary-tree-level-order-traversal/)
//! - [#994 Rotting Oranges](https://leetcode.com/problems/rotting-oranges/)
//!
//! ## Use Cases
//!
//! - BFS (Breadth-First Search)
//! - Task scheduling
//! - Print job spooling
//! - Message queues
//! - Request handling (web servers)
//! - Buffering (I/O operations)
//!
//! ## Example
//!
//! ```rust
//! use dsa_data_structures::linear::Queue;
//!
//! let mut queue = Queue::new();
//! queue.enqueue(1);
//! queue.enqueue(2);
//! queue.enqueue(3);
//!
//! assert_eq!(queue.len(), 3);
//! assert_eq!(queue.peek(), Some(&1));
//! assert_eq!(queue.dequeue(), Some(1));
//! assert_eq!(queue.dequeue(), Some(2));
//! ```

use alloc::collections::VecDeque;
use alloc::vec::Vec;

use dsa_core::{Container, Searchable};

/// A FIFO (First In, First Out) queue implementation.
///
/// This implementation uses `VecDeque` as the underlying storage,
/// providing amortized O(1) operations at both ends.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Queue<T> {
    data: VecDeque<T>,
}

impl<T> Queue<T> {
    /// Creates a new empty queue.
    ///
    /// # Time Complexity
    /// O(1)
    ///
    /// # Example
    ///
    /// ```rust
    /// use dsa_data_structures::linear::Queue;
    ///
    /// let queue: Queue<i32> = Queue::new();
    /// assert!(queue.is_empty());
    /// ```
    #[must_use]
    pub fn new() -> Self {
        Queue {
            data: VecDeque::new(),
        }
    }

    /// Creates a new queue with the specified capacity.
    ///
    /// # Time Complexity
    /// O(1)
    #[must_use]
    pub fn with_capacity(capacity: usize) -> Self {
        Queue {
            data: VecDeque::with_capacity(capacity),
        }
    }

    /// Returns the number of elements in the queue.
    ///
    /// # Time Complexity
    /// O(1)
    #[inline]
    #[must_use]
    pub fn len(&self) -> usize {
        self.data.len()
    }

    /// Returns `true` if the queue contains no elements.
    ///
    /// # Time Complexity
    /// O(1)
    #[inline]
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.data.is_empty()
    }

    /// Adds an element to the back of the queue.
    ///
    /// # Time Complexity
    /// O(1) amortized
    ///
    /// # Example
    ///
    /// ```rust
    /// use dsa_data_structures::linear::Queue;
    ///
    /// let mut queue = Queue::new();
    /// queue.enqueue(10);
    /// queue.enqueue(20);
    /// assert_eq!(queue.peek(), Some(&10));
    /// ```
    pub fn enqueue(&mut self, value: T) {
        self.data.push_back(value);
    }

    /// Removes and returns the front element from the queue.
    ///
    /// Returns `None` if the queue is empty.
    ///
    /// # Time Complexity
    /// O(1)
    ///
    /// # Example
    ///
    /// ```rust
    /// use dsa_data_structures::linear::Queue;
    ///
    /// let mut queue = Queue::new();
    /// queue.enqueue(10);
    /// queue.enqueue(20);
    /// assert_eq!(queue.dequeue(), Some(10));
    /// assert_eq!(queue.dequeue(), Some(20));
    /// assert_eq!(queue.dequeue(), None);
    /// ```
    pub fn dequeue(&mut self) -> Option<T> {
        self.data.pop_front()
    }

    /// Returns a reference to the front element without removing it.
    ///
    /// Returns `None` if the queue is empty.
    ///
    /// # Time Complexity
    /// O(1)
    ///
    /// # Example
    ///
    /// ```rust
    /// use dsa_data_structures::linear::Queue;
    ///
    /// let mut queue = Queue::new();
    /// queue.enqueue(10);
    /// assert_eq!(queue.peek(), Some(&10));
    /// assert_eq!(queue.len(), 1); // Element not removed
    /// ```
    #[must_use]
    pub fn peek(&self) -> Option<&T> {
        self.data.front()
    }

    /// Returns a mutable reference to the front element without removing it.
    ///
    /// # Time Complexity
    /// O(1)
    #[must_use]
    pub fn peek_mut(&mut self) -> Option<&mut T> {
        self.data.front_mut()
    }

    /// Returns a reference to the back element.
    ///
    /// # Time Complexity
    /// O(1)
    #[must_use]
    pub fn back(&self) -> Option<&T> {
        self.data.back()
    }

    /// Clears the queue, removing all elements.
    ///
    /// # Time Complexity
    /// O(n)
    pub fn clear(&mut self) {
        self.data.clear();
    }

    /// Returns an iterator over the queue from front to back.
    pub fn iter(&self) -> impl Iterator<Item = &T> {
        self.data.iter()
    }

    /// Converts the queue to a `Vec` (front to back).
    #[must_use]
    pub fn to_vec(&self) -> Vec<T>
    where
        T: Clone,
    {
        self.data.iter().cloned().collect()
    }

    /// Creates a queue from a `Vec` (first element is front).
    #[must_use]
    pub fn from_vec(vec: Vec<T>) -> Self {
        Queue {
            data: vec.into_iter().collect(),
        }
    }
}

impl<T: PartialEq> Queue<T> {
    /// Returns `true` if the queue contains the specified value.
    ///
    /// # Time Complexity
    /// O(n)
    #[must_use]
    pub fn contains(&self, value: &T) -> bool {
        self.data.contains(value)
    }
}

impl<T> Container for Queue<T> {
    fn len(&self) -> usize {
        self.data.len()
    }
}

impl<T: PartialEq> Searchable<T> for Queue<T> {
    fn search(&self, value: &T) -> Option<usize> {
        self.data.iter().position(|x| x == value)
    }
}

impl<T> Default for Queue<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T> FromIterator<T> for Queue<T> {
    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
        Queue {
            data: iter.into_iter().collect(),
        }
    }
}

impl<T> IntoIterator for Queue<T> {
    type Item = T;
    type IntoIter = alloc::collections::vec_deque::IntoIter<T>;

    fn into_iter(self) -> Self::IntoIter {
        self.data.into_iter()
    }
}

impl<'a, T> IntoIterator for &'a Queue<T> {
    type Item = &'a T;
    type IntoIter = alloc::collections::vec_deque::Iter<'a, T>;

    fn into_iter(self) -> Self::IntoIter {
        self.data.iter()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    mod basics {
        use super::*;

        #[test]
        fn test_new() {
            let queue: Queue<i32> = Queue::new();
            assert!(queue.is_empty());
            assert_eq!(queue.len(), 0);
        }

        #[test]
        fn test_with_capacity() {
            let queue: Queue<i32> = Queue::with_capacity(10);
            assert!(queue.is_empty());
        }

        #[test]
        fn test_default() {
            let queue: Queue<i32> = Queue::default();
            assert!(queue.is_empty());
        }

        #[test]
        fn test_from_vec() {
            let queue = Queue::from_vec(vec![1, 2, 3]);
            assert_eq!(queue.len(), 3);
            assert_eq!(queue.peek(), Some(&1));
        }

        #[test]
        fn test_to_vec() {
            let queue = Queue::from_vec(vec![1, 2, 3]);
            assert_eq!(queue.to_vec(), vec![1, 2, 3]);
        }

        #[test]
        fn test_clone() {
            let queue = Queue::from_vec(vec![1, 2, 3]);
            let cloned = queue.clone();
            assert_eq!(queue, cloned);
        }
    }

    mod enqueue_dequeue {
        use super::*;

        #[test]
        fn test_enqueue() {
            let mut queue = Queue::new();
            queue.enqueue(1);
            queue.enqueue(2);
            queue.enqueue(3);
            assert_eq!(queue.len(), 3);
            assert_eq!(queue.peek(), Some(&1));
        }

        #[test]
        fn test_dequeue() {
            let mut queue = Queue::from_vec(vec![1, 2, 3]);
            assert_eq!(queue.dequeue(), Some(1));
            assert_eq!(queue.dequeue(), Some(2));
            assert_eq!(queue.dequeue(), Some(3));
            assert_eq!(queue.dequeue(), None);
        }

        #[test]
        fn test_fifo_order() {
            let mut queue = Queue::new();
            queue.enqueue(1);
            queue.enqueue(2);
            queue.enqueue(3);

            // Should come out in same order (FIFO)
            assert_eq!(queue.dequeue(), Some(1));
            assert_eq!(queue.dequeue(), Some(2));
            assert_eq!(queue.dequeue(), Some(3));
        }
    }

    mod peek {
        use super::*;

        #[test]
        fn test_peek() {
            let mut queue = Queue::new();
            assert_eq!(queue.peek(), None);

            queue.enqueue(10);
            assert_eq!(queue.peek(), Some(&10));
            assert_eq!(queue.len(), 1); // Not removed

            queue.enqueue(20);
            assert_eq!(queue.peek(), Some(&10)); // Still front element
        }

        #[test]
        fn test_peek_mut() {
            let mut queue = Queue::from_vec(vec![1, 2, 3]);
            if let Some(front) = queue.peek_mut() {
                *front = 10;
            }
            assert_eq!(queue.peek(), Some(&10));
        }

        #[test]
        fn test_back() {
            let queue = Queue::from_vec(vec![1, 2, 3]);
            assert_eq!(queue.back(), Some(&3));
        }
    }

    mod search {
        use super::*;

        #[test]
        fn test_contains() {
            let queue = Queue::from_vec(vec![1, 2, 3]);
            assert!(queue.contains(&2));
            assert!(!queue.contains(&5));
        }
    }

    mod utilities {
        use super::*;

        #[test]
        fn test_clear() {
            let mut queue = Queue::from_vec(vec![1, 2, 3]);
            queue.clear();
            assert!(queue.is_empty());
        }

        #[test]
        fn test_iter() {
            let queue = Queue::from_vec(vec![1, 2, 3]);
            let collected: Vec<_> = queue.iter().cloned().collect();
            assert_eq!(collected, vec![1, 2, 3]);
        }

        #[test]
        fn test_from_iter() {
            let queue: Queue<i32> = (1..=5).collect();
            assert_eq!(queue.len(), 5);
            assert_eq!(queue.peek(), Some(&1));
        }

        #[test]
        fn test_into_iter() {
            let queue = Queue::from_vec(vec![1, 2, 3]);
            let collected: Vec<_> = queue.into_iter().collect();
            assert_eq!(collected, vec![1, 2, 3]);
        }
    }

    mod edge_cases {
        use super::*;

        #[test]
        fn test_empty_operations() {
            let mut queue: Queue<i32> = Queue::new();
            assert_eq!(queue.dequeue(), None);
            assert_eq!(queue.peek(), None);
            assert!(!queue.contains(&1));
        }

        #[test]
        fn test_single_element() {
            let mut queue = Queue::new();
            queue.enqueue(42);
            assert_eq!(queue.peek(), Some(&42));
            assert_eq!(queue.dequeue(), Some(42));
            assert!(queue.is_empty());
        }

        #[test]
        fn test_large_queue() {
            let mut queue = Queue::new();
            for i in 0..10000 {
                queue.enqueue(i);
            }
            assert_eq!(queue.len(), 10000);
            assert_eq!(queue.peek(), Some(&0));
        }

        #[test]
        fn test_interleaved_operations() {
            let mut queue = Queue::new();
            queue.enqueue(1);
            queue.enqueue(2);
            assert_eq!(queue.dequeue(), Some(1));
            queue.enqueue(3);
            assert_eq!(queue.dequeue(), Some(2));
            assert_eq!(queue.dequeue(), Some(3));
            assert!(queue.is_empty());
        }
    }

    mod equality {
        use super::*;

        #[test]
        fn test_eq() {
            let q1 = Queue::from_vec(vec![1, 2, 3]);
            let q2 = Queue::from_vec(vec![1, 2, 3]);
            let q3 = Queue::from_vec(vec![1, 2, 4]);
            assert_eq!(q1, q2);
            assert_ne!(q1, q3);
        }
    }
}
